### DISCUSSION TITLE
Ingress gateway 404 on TLS routing

### DISCUSSION THREAD

**Discussion by elysweyr** [istio/istio#52452] [Q&A] [answered]
Hi,

**Outline:** My ingress router responds with a 404 on TLS traffic (HTTP traffic works fine!) and I don't know why. 
Envoy seems to mark this traffic with `transport socket match, socket tlsMode-disabled` which seems highly suspicious to me.

I'm writing here as I'm working already a few days on this issue and didn't figure out how to completely fix it, yet. What I'm trying to achieve is to add Keycloak to a service mesh and use reference this Keycloak instance in a RequestAuthentication CR of the same service mesh. 

1. **Keycloak is up and running and externally reachable** (https://sso.apps.reavo.dev/)
   This means that:
   * K8s ingress is configured and working
   * Service mesh's ingress controller is running
   * [Gateway object](https://gist.github.com/elysweyr/a1893a7ede60cd1eca0e83bef6e30674) has been configured (80, 443)
   * [Ingress configuration](https://gist.github.com/elysweyr/5ca688f74c93c235fb3d8d0fa2cee9aa) exists with `tls.termination` set to `edge`, `host` set to `sso.apps.reavo.dev` and pointing to the ingress gateway's  service on port 80 (Had the same overall issue when trying to route this via TLS)
   * [Virtual service](https://gist.github.com/elysweyr/cdc8e37c35b17d5fefd31a9cb68d9569) has been configured
2. **Configuration for enabling connectivity using external keycloak dns name `sso.apps.reavo.dev` has been configured**
    * [Service entry](https://gist.github.com/elysweyr/736085fc11e20ac5fc6b397085c35ec7) exists
3. **`istioctl analyze` has executed on all relevant namespaces**

```
❯ istioctl analyze -n reavo-prod
Info [IST0118] (Service reavo-prod/reavo-keycloak-discovery) Port name  (port: 7800, targetPort: 7800) doesn't follow the naming convention of Istio port. # This port is used for Keycloak clustering which isn't relevant for routing traffic to this application
Info [IST0118] (Service reavo-prod/reavo-sso-pg-ha) Port name postgres (port: 5432, targetPort: postgres) doesn't follow the naming convention of Istio port. # I'm aware of this misconfiguration but postgres isn't part of this scope yet as I want to set up a working TLS routing configuration before going any deeper
Info [IST0118] (Service reavo-prod/reavo-sso-pg-primary) Port name postgres (port: 5432, targetPort: postgres) doesn't follow the naming convention of Istio port. # Same as above
Info [IST0118] (Service reavo-prod/reavo-sso-pg-replicas) Port name postgres (port: 5432, targetPort: postgres) doesn't follow the naming convention of Istio port. # Same as above

❯ istioctl analyze -n reavo-ingress-gateway

✔ No validation issues found when analyzing namespace: reavo-ingress-gateway.

```

4. **Testing TLS connectivity from different sources**
  * Local pc - ✅Working
  ```
  curl https://sso.apps.reavo.dev/ -I
  HTTP/1.1 302 Found
   ```
  * K8s pod (not part of any mesh) - ❌Not working
  ```
# Testing HTTPS
sh-4.4$ curl https://sso.apps.reavo.dev -I
HTTP/2 404 
date: Wed, 31 Jul 2024 21:56:51 GMT
server: istio-envoy

# HTTPS isn't working so now I'm switching to HTTP
sh-4.4$ curl http://sso.apps.reavo.dev -I
HTTP/1.1 302 Found
[...]
server: istio-envoy
[...]
   ```
  * K8s envoy sidecar (part of service mesh in question; used for further debugging) - ❌Not working
  ```
# Testing HTTPS
sh-4.4$ curl https://sso.apps.reavo.dev -I
HTTP/2 404 
date: Wed, 31 Jul 2024 21:58:30 GMT
server: istio-envoy

# HTTPS isn't working so now I'm switching to HTTP
sh-4.4$ curl http://sso.apps.reavo.dev -I
HTTP/1.1 302 Found
[...]
server: istio-envoy
[...]
   ```
**5. Set sidecar log level to debug**
```
istioctl proxy-config log hello-openshift-57f6fff586-g6mdg.reavo-test --level debug
```
**6. Retried curl and fetched logs**
After analyzing the [logs](https://gist.github.com/elysweyr/c7ffd722b5a0c5568db6c2e4e9431593) I've found these interesting log lines
```
2024-07-31T21:11:03.625686Z	debug	envoy dns external/envoy/source/extensions/network/dns_resolver/cares/dns_impl.cc:275	dns resolution for reavo-keycloak-service.reavo-prod.svc.cluster.local completed with status 0	thread=19
2024-07-31T21:11:03.625710Z	debug	envoy upstream external/envoy/source/common/upstream/upstream_impl.cc:457	transport socket match, socket tlsMode-disabled selected for host with address 172.30.158.98:8443	thread=19
2024-07-31T21:11:03.625735Z	debug	envoy upstream external/envoy/source/extensions/clusters/strict_dns/strict_dns_cluster.cc:177	DNS refresh rate reset for reavo-keycloak-service.reavo-prod.svc.cluster.local, refresh rate 5000 ms	thread=19
```

`172.30.158.98` is Keycloak's service IP address and 8443 is the TLS port specified in the `ServiceEntry` CR
Could this be the beginnig of finding a solution to my problem?

I'm very grateful for any hints you have! 
Thanks in advance!

**Comment by howardjohn** [comment, marked as answer]
```
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: sso-virtual-service
  namespace: reavo-prod
spec:
  gateways:
    - reavo-ingress-gateway/ingressgateway
  hosts:
    - sso.apps.reavo.dev
  http:
    - match:
        - port: 80
      route:
        - destination:
            host: reavo-keycloak-service.reavo-prod.svc.cluster.local
            port:
              number: 8080
  tls:
    - match:
        - sniHosts:
            - sso.apps.reavo.dev
      route:
        - destination:
            host: reavo-keycloak-service.reavo-prod.svc.cluster.local
            port:
              number: 8443
```

This is the problem. You should not be doing a TLS match here. That is used to match **unterminated TLS**. Instead you just want to do an `http` match - this works for the HTTPS traffic since it works after the TLS is terminated.

Example: https://istio.io/latest/docs/tasks/traffic-management/ingress/secure-ingress/#configure-a-tls-ingress-gateway-for-a-single-host

**Reply by elysweyr** [reply]
Yes, it seems that I misunderstood the TLS matching concept here. Removing the TLS and the explicit `port: 80` match has had the desired effect.

Thanks @howardjohn for your quick and helpful answer!
